rest-api/src/main/scala/za/co/absa/enceladus/rest_api/services/v3/VersionedModelServiceV3.scala

```diff
@@ -0,0 +1,111 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.rest_api.services.v3
+
+import org.mongodb.scala.result.UpdateResult
+import org.slf4j.LoggerFactory
+import org.springframework.security.core.context.SecurityContextHolder
+import org.springframework.security.core.userdetails.UserDetails
+import za.co.absa.enceladus.model.{UsedIn, Validation}
+import za.co.absa.enceladus.model.menas.audit._
+import za.co.absa.enceladus.model.versionedModel.{VersionedModel, VersionedSummary}
+import za.co.absa.enceladus.rest_api.exceptions._
+import za.co.absa.enceladus.rest_api.services.VersionedModelService
+
+import scala.concurrent.Future
+
+// scalastyle:off number.of.methods
+trait VersionedModelServiceV3[C <: VersionedModel with Product with Auditable[C]] extends VersionedModelService[C] {
+
+  import scala.concurrent.ExecutionContext.Implicits.global
+
+  override private[services] val logger = LoggerFactory.getLogger(this.getClass)
+
+  def getLatestVersionSummary(name: String): Future[Option[VersionedSummary]] = {
+    mongoRepository.getLatestVersionSummary(name)
+  }
+
+  // v3 has internal validation on importItem (because it is based on update), v2 only had it on import
+  override def importSingleItem(item: C, username: String, metadata: Map[String, String]): Future[Option[(C, Validation)]] = {
+    val metadataValidation = validateMetadata(metadata)
+    if (metadataValidation.isValid) {
+      // even if valid, merge validations results (warnings may be theoretically present)
+      importItem(item, username).map(_.map { case (item, validation) => (item, validation.merge(metadataValidation)) })
+    } else {
+      Future.failed(ValidationException(metadataValidation))
+    }
+  }
+
+  override private[services] def update(username: String, itemName: String, itemVersion: Int)
+                                       (transform: C => C): Future[Option[(C, Validation)]] = {
+    this.updateFuture(username, itemName, itemVersion) { item: C =>
+      for {
+        _ <- validate(item) // V3 validates the to-be-updated item, too (V2 does not)
+       updatedItem <- Future {
+        transform(item)
+        }
+      } yield updatedItem
+    }
+  }
+
+  /**
+   * Enables all versions of the entity by name.
+   * @param name
+   */
+  def enableEntity(name: String): Future[UpdateResult] = {
+    val auth = SecurityContextHolder.getContext.getAuthentication
+    val principal = auth.getPrincipal.asInstanceOf[UserDetails]
+
+    // todo check validation of used-in dependees? #2065
```

Comment:
I think this comment can be deleted

Reply:
No, we cannot, because feature #2065 is still missing and it should be implemented here. But currently, it has not been given priority.